rpcs3/Emu/RSX/GL/rsx_gl_texture.cpp

```diff
@@ -226,60 +295,39 @@ namespace rsx
 
 			case CELL_GCM_TEXTURE_R6G5B5:
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 2);
-
-				// TODO: Probably need to actually unswizzle if is_swizzled.
-				const u32 numPixels = tex.width() * tex.height();
-				unswizzledPixels = (u8 *)malloc(numPixels * 4);
-				// TODO: Speed.
-				for (u32 i = 0; i < numPixels; ++i) {
-					u16 c = reinterpret_cast<const be_t<u16> *>(pixels)[i];
-					unswizzledPixels[i * 4 + 0] = convert_6_to_8((c >> 10) & 0x3F);
-					unswizzledPixels[i * 4 + 1] = convert_5_to_8((c >> 5) & 0x1F);
-					unswizzledPixels[i * 4 + 2] = convert_5_to_8((c >> 0) & 0x1F);
-					unswizzledPixels[i * 4 + 3] = 255;
-				}
-
-				glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, tex.width(), tex.height(), 0, GL_RGBA, GL_UNSIGNED_BYTE, unswizzledPixels);
-
-				free(unswizzledPixels);
+				glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, tex.width(), tex.height(), 0, GL_RGBA, GL_UNSIGNED_BYTE, texture_data);
 				break;
 			}
 
 			case CELL_GCM_TEXTURE_DEPTH24_D8: //  24-bit unsigned fixed-point number and 8 bits of garbage
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 4);
-
-				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_BYTE, pixels);
+				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_BYTE, texture_data);
 				break;
 			}
 
 			case CELL_GCM_TEXTURE_DEPTH24_D8_FLOAT: // 24-bit unsigned float and 8 bits of garbage
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 4);
-				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_FLOAT, pixels);
+				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_FLOAT, texture_data);
 				break;
 			}
 
 			case CELL_GCM_TEXTURE_DEPTH16: // 16-bit unsigned fixed-point number
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 2);
-				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT16, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_SHORT, pixels);
+				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT16, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_SHORT, texture_data);
 				break;
 			}
 
 			case CELL_GCM_TEXTURE_DEPTH16_FLOAT: // 16-bit unsigned float
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 2);
-				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT16, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_FLOAT, pixels);
+				glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT16, tex.width(), tex.height(), 0, GL_DEPTH_COMPONENT, GL_FLOAT, texture_data);
 				break;
 			}
 
 			case CELL_GCM_TEXTURE_X16: // A 16-bit fixed-point number
 			{
-				glPixelStorei(GL_UNPACK_ROW_LENGTH, pitch / 2);
+				LOG_WARNING(RSX, "CELL_GCM_X16 texture. Watch out for corruption due to swapped color channels!");
```

Comment:
@danilaml There may be errors due to swapping of byte order, especially if the texture is supplied swizzled. This makes some textures require alignment/expansion regardless of the input to standardize behaviour. I observed this with the A1R5G5B5 format and thats why it is always aligned. This however can cause significant slowdowns if overused.